### PR TITLE
Update soon to be expired crl

### LINKED ISSUE #2357: Tests fail in 2019-11-25
While working towards openSUSE reproducible builds, I found that mbedtls does not build in 2020

### Description
- Type: Bug
- Priority: Minor

---------------------------------------------------------------
## Bug

**mbed TLS build:**  
Version: 2.14.1
OS version: any

**Expected behavior** 
software should build anytime anywhere

**Actual behavior**  
`FAILED (486 / 518 tests (86 skipped))`
See full log in
https://bugzilla.opensuse.org/show_bug.cgi?id=1081871#c6

**Steps to reproduce**  
build mbedtls in 2020 and run its testsuite
e.g. on openSUSE or Debian
have a free openSUSE account,
have `kvm`, `osc` and `build` or `obs-build` packages installed. Then run
```bash
osc checkout openSUSE:Factory/mbedtls && cd $_
osc build --vm-type=kvm --build-opt=--vm-custom-opt="-rtc base=2020-03-01T01:01:01" --noservice
```

Could probably be solved similarly to https://dev.gnupg.org/T3815
with a cert that never expires (or only in the year 2999)

### PR DESCRIPTION
**Pull request by RonEld**
## Description
Update crl.pem, as it will expire on November 25 2019.
Resolves #2357.

## Status
**READY**

## Requires Backporting
Yes

As this crl will be expired on the LTS branch, which will still be supported on November 25 2019,  need to backport as well.

## Todos
- [x] Tests
- [ ] Documentation
- [ ] Changelog updated
- [ ] Backported

### CODE REVIEW

**Comment by RonEld** [comment]
@mpg Thank you for your review!

I have just noticed I have missed adding `crl.pem` as a dependency in the Makefile.

Backported to 2.7 and 2.16 in #2418 and #2419

**Comment by RonEld** [comment]
@hanno-arm Yes. All the other CRLs we have, the soonest year they expire is 2024.

We have other CRLs that expired at 2011, so since we didn't get any errors sine, they are probably not used in our tests, and we require some house cleaning for certificates and crls not used in tests.

**Comment by mpg** [comment]
Or perhaps the expired CRLs are used in tests that actually need the CRL to be expired?

**Comment by RonEld** [comment]
perhaps.
Anyway, when I run:
`faketime -f '+2y' make test` , the tests pass. ( 3 years ahead I get failures.

I have noticed that: `data_files/test-ca.crt` is valid until:
~~~
Not After : Feb 12 14:44:00 2021 GMT
~~~
and that `data_files/server3.crt` is valid until:
~~~
Not After : Aug  7 09:17:03 2023 GMT
~~~

I haven't checked other certificates. Anyway, this should be handled in a different PR, and maintained regularly.

**Comment by hanno-arm** [comment]
@RonEld I'd prefer those CRTs to be regenerated (and their build-instructions to be rewritten) so that they don't expire in the next 3 years. Now you/we are into the issue and it's easy to fix along the way, while waiting for another report to come up in a year's time or so and triaging and fixing it again will be more work.

**Comment by mpg** [comment]
In addition to what Hanno wrote, our LTS branches are supported for 3 years, so I think it would make sense for us to always ensure `faketime -f '+3y' make test` pass when we release a new LTS (and released one very recently, so I think we should have it pass now).

**Comment by RonEld** [comment]
@hanno-arm @mpg I agree that we should have a test that will ensure `faketime -f '+3y' make test` will pass. However, I disagree that this should be handled in this PR, as it is an entirely different task, as I think we should align all our certificates and crls to be expired on the same time, for simplicity

**Comment by mpg** [comment]
@RonEld I'm not sure this is an entirely different task - this PR is about making tests pass in 1 year, so just changing the numeric value from 1 to 3 hardly changes the nature of the task IMO. (The larger work you suggest in your email would indeed be a pretty different task.)

However, if you really to handle the two remaining certs necessary to pass in 3 years in this PR, I'll still approve this PR as it stands in the interest of unblocking it, as it already brings a valuable improvement that I don't want to block by requesting more of the same.

**Comment by RonEld** [comment]
I am working on making the tests pass in three years. however htis is more than just changing two certificates

**Comment by mpg** [comment]
@RonEld Why is that more than just changing the two certificates you mentioned in [this earlier message](https://github.com/ARMmbed/mbedtls/pull/2417#issuecomment-462258255)? How much more?

I based my previous comment on the assumption that it would be just changing two certs, so if it's much more that than I could change my mind.

**Comment by RonEld** [comment]
The task for making it poass in three years include:
1. As you can see in the list I stated in my email, you will see many certificates that expire in three years. ( I listed the main server1.crt and test-ca.crt but all their derivatives as well). THis is not a big issue though.
2. Changing the Makefile in data_files to generate update certificates. Not a big issue.
3. Modifying the .data file in the x509parse test suite, to have the correc tinformation. This is not difficult, only takes time.

As mentioned, I am currently working on this, probably be ready by end of day

**Comment by RonEld** [comment]
@hanno-arm @mpg I have updated the certificates and corresponding test data files to make our tests pass 3 years ahead.
I still think this is a patch, as it should be handled in a more maintained way though.

I ran the following tests to verify:
~~~
faketime -f '+3y' make test
faketime -f '+3y' tests/ssl-opts.sh
faketime -f '+3y' programs/ssl/ssl_server2 and faketime -f '+3y' programs/ssl/ssl_client2
~~~

A few notes:
* `server2-badsign.crt` was generated manually, using a debugger to corrupt the signature, so no Makefile entry done for that.
* `enco-ca-prstr.pem` was generated through `mbedtls_cert_write` program, since openssl didn't generate the CN in printablestring. Our certificate write writes in UTF8, so manually modified this. Therefore no Makefile entry for that as well.
* I am not sure what is the purpose of `test-ca_uppercase.crt`, as the content of original file was not in uppercase. I generated the file, as original CN.

Please review

**Comment by RonEld** [comment]
Note: I think it's worth fixing #822 as part of this PR

**Comment by bmwiedemann** [comment]
patch applies cleanly to 2.16.0 but only adds another 10 years to the life of the software:
```
openssl crl -in ./tests/data_files/crl.pem -noout -text
        Last Update: Feb  6 16:34:46 2019 GMT
        Next Update: Feb  6 16:34:46 2029 GMT
```

cant you make it expire in the year 2999?

**Comment by AndrzejKurek** [comment]
@mpg, @hanno-arm - please re-review this PR.

**Comment by RonEld** [comment]
After think a bit some more, #822 should be handled in a different PR,as part of a house cleaning task, removing unused certificates.

**Comment by RonEld** [comment]
@bmwiedemann 
Yes, the crls are valid for 10 years, as they were before.Sine the lifespan of our LTS branches are ~3 years, and these are test certificates, I think its reasonable to have these certificates expire in 10 years.
However I am willing to hear other thoughts on this matter.

@mpg @hanno-arm any comments?

**Comment by RonEld** [comment]
Keeping the "needs work" label until conflicts are resolved.

**Comment by RonEld** [comment]
I have rebased to resolve conflixts.
I have also added a commit to update expiration date of new certificates.

@mpg @hanno-arm Do you want me to squash https://github.com/ARMmbed/mbedtls/pull/2417/commits/66d99ae2aa9343b4d5e74f5ba312676595874168 with https://github.com/ARMmbed/mbedtls/pull/2417/commits/23054c8b9480fb94428913d4361aea2ebf9e4b5a?

**Comment by mpg** [comment]
> @mpg @hanno-arm Do you want me to squash 66d99ae with 23054c8?

As you prefer, both are fine with me - though I have a slight preference for squashing if nobody opposes.

Also, github is still showing conflicts.

**Comment by RonEld** [comment]
@mpg yes, the new conflict is from the merging of #1520 
I will rebase to fix the merge conflict.

>As you prefer, both are fine with me - though I have a slight preference for squashing if nobody opposes.

Personally I think it's better if I squashed the commits, but I think it will make the reviewers an easier life if the squashing will happen after all the changes will be reviewed

**Comment by RonEld** [comment]
@mpg I have resolved the conflicts. Please review.
I have noticed  a typo in the last commit message, but I wouldn't bother, since the last two commits will be squashed to https://github.com/ARMmbed/mbedtls/pull/2417/commits/e87c6855bd7601cf8c7c5063990a8068c42b3fc2

**Comment by mpg** [comment]
Thanks! Actually, would you be able to find another reviewer? Atm as you now I'm fully dedicated to some other work.

**Comment by RonEld** [comment]
@mpg sure, no problem.
I just thought that since you started this, it would be faster if you finish.
I'll assign other people

File: tests/data_files/cert_example_wildcard.crt [hunk 1]
```diff
@@ -1,77 +1,20 @@
-Certificate:
```
Comment (AndrzejKurek) [inline tests/data_files/cert_example_wildcard.crt]: Don't we want to have such information about the new certificate here?

Comment (RonEld) [reply tests/data_files/cert_example_wildcard.crt]: We can, but what's the added value?
Frankly I don't think this should block the PR.

**Comment by RonEld** [comment]
@AndrzejKurek Thank you for the review.
As mentioned in https://github.com/ARMmbed/mbedtls/pull/2417#issuecomment-470439682, I have squashed the latest commit to clear history, but you can look in [this diff](https://github.com/ARMmbed/mbedtls/compare/b40b4bb36ef19a9efc29a04dfdaf9e730400cb9b..e38850a8917608b1efec34ea270cd071de1c7591) that nothing has changed in the process since your last approval.
I apologize for not clearing the history prior to your review.

**Comment by RonEld** [comment]
@k-stachowiak @AndrzejKurek I amended the commit "Update certificates to expire in 2029 " to fix CI failure.
It was only removing of an extra `;` character in `certs.c` line 203. This caused a failure in ARMCC5:
~~~
"certs.c", line 205: Error:  #381-D: extra ";" ignored
  const char mbedtls_test_srv_crt_rsa[]     =  
EST_SRV_CRT_RSA_SHA256;
                                                                      ^
certs.c: 0 warnings, 1 error
Makefile:164: recipe for target 'certs.o' failed
make[1]: *** [certs.o] Error 1
make[1]: *** Waiting for unfinished jobs....
make: *** [lib] Error 2
Makefile:18: recipe for target 'lib' failed
^^^^build_armcc: build: ARM Compiler 5, make: command make CC=/usr/local/ARM_Compiler_5.06u3/bin//armcc AR=/usr/local/ARM_Compiler_5.06u3/bin//armar WARNING_CFLAGS=--strict --c99 lib -> 2^^^^
~~~
Please review

**Comment by RonEld** [comment]
This PR and its backports have been fully reviewed and approved. Removing the "needs backports" label

cc @Patater

**Comment by bmwiedemann** [comment]
Hi, I'm wondering what is missing here for this to get merged? The 2019-11-25 expiry of the crl is not that far away.

**Comment by bmwiedemann** [comment]
collides with
```
commit 53756b32282d0b36a6900cf2459250a328ae6214
Author: Hanno Becker <...@arm.com>
Date:   Mon Jun 3 14:14:38 2019 +0100

    Add MD[245] test CRTs to tree
```

**Comment by RonEld** [comment]
@k-stachowiak @AndrzejKurek I have rebased to resolve conflicts
During rebasing, I found new certificates that need to be updated, and other certificates that fail with `faketime -f '+3y'` since this PR was raised.
Please review

**Comment by RonEld** [comment]
This PR is fully reviewed, so removing the "needs backports" label
cc @Patater

**Pull request merged**